### PR TITLE
chore: add correct typings and fix other warnings reported by ESLint

### PR DESCRIPTION
**Pull request by hohy**
## Changes
I replaced all usage of the `any` type in the code with correct typings and fix all other warnings reported by the ESLint.

I tried to improve mocking in the project. I removed a lot of unnecessary mock from unit tests and replaced them with global mocks from the `__mocks__` folder. But there is still opportunity for improvement. I would like to remove mocks like this one `vscode.window.showErrorMessage = jest.fn();`. I think we should avoid this kind of mocks as they are not restored by `jest.clearAllMocks();`, so I suggest to replace them with `jest.spyOn`. But I would do that in separate PR as this one is already quite big.

There should be no changes in business logic, only exception is in the `Settings` service where I refactored the `lspConfigHandler` function.

Sorry for a big PR 😞 

## How Has This Been Tested?

Unit and integration tests

## Checklist:
- [x] Each of my commits contains one meaningful change
- [x] I have performed rebase of my branch on top of the development
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

**Comment by hohy** [comment]
@slavek-kucera @VitGottwald ready for review. Thanks!

File: clients/cobol-lsp-vscode-extension/src/__tests__/commands/SmartTabCommandTest.spec.ts [hunk 1]
```diff
@@ -12,89 +12,95 @@
...
     new smartTab.SmartTabCommandProvider(context, "some name");
   const smartOut: smartTab.SmartTabCommandProvider =
     new smartTab.SmartOutdentCommandProvider(context, "some name");
   const active: Position = new Position(1, 2);
-  const mockSelection: Selection = new Selection(active, active);
-  (mockSelection as any).start = { line: 1 };
-  (mockSelection as any).end = { line: 3 };
+  const mockSelection = new Selection(active, active);
+  asMutable(mockSelection).start = new Position(1, 1);
+  asMutable(mockSelection).end = new Position(3, 1);
```
Comment (slavek-kucera) [inline clients/cobol-lsp-vscode-extension/src/__tests__/commands/SmartTabCommandTest.spec.ts:70]: Why was the column selected as `1`?

File: clients/cobol-lsp-vscode-extension/src/__tests__/services/LanguageClientServiceTest.spec.ts [hunk 2]
```diff
@@ -249,55 +229,55 @@ describe("LanguageClientService positive scenario", () => {
...
-    ).executableService.getNativeLanguageClient();
-    expect(executable.command).toBe("engine.exe");
-    expect(executable.options.cwd).toBe(join("/test", "native"));
+    languageClientService["executableService"] = new NativeExecutableService(
+      "/test",
+    );
+    const executable =
+      languageClientService["executableService"].getNativeLanguageClient();
+    expect(executable?.command).toBe("engine.exe");
+    expect(executable?.options?.cwd).toBe(join("/test", "native"));
```
Comment (slavek-kucera) [inline clients/cobol-lsp-vscode-extension/src/__tests__/services/LanguageClientServiceTest.spec.ts]: These are not exactly realistic windows paths.

File: clients/cobol-lsp-vscode-extension/src/__tests__/services/snippets/SnippetTest.spec.ts [hunk 3]
```diff
@@ -274,9 +254,9 @@ describe("Test CompletionProvider", () => {
     });
     const completions = await snippetcompletion.provideCompletionItems(
       doc,
-      position as any,
+      new vscode.Position(0, 5),
```
Comment (slavek-kucera) [inline clients/cobol-lsp-vscode-extension/src/__tests__/services/snippets/SnippetTest.spec.ts:257]: why `4`->`5`?

File: clients/cobol-lsp-vscode-extension/src/__tests__/services/snippets/SnippetTest.spec.ts [hunk 4]
```diff
@@ -247,9 +230,9 @@ describe("Test CompletionProvider", () => {
       (
         await snippetcompletion.provideCompletionItems(
           doc,
-          position as any,
+          new vscode.Position(0, 5),
```
Comment (slavek-kucera) [inline clients/cobol-lsp-vscode-extension/src/__tests__/services/snippets/SnippetTest.spec.ts:233]: also here...

File: clients/cobol-lsp-vscode-extension/src/commands/RunAnalysisCLI.ts [hunk 5]
```diff
@@ -81,7 +81,7 @@ export class RunAnalysis {
   /**
    * Prompt the user for the location of the copybook config file.
    */
-  public async getCopybookConfigLocation(result: Partial<AnalysisResults>) {
+  public getCopybookConfigLocation(result: Partial<AnalysisResults>) {
```
Comment (slavek-kucera) [inline clients/cobol-lsp-vscode-extension/src/commands/RunAnalysisCLI.ts]: Could you perhaps de-Java-fy this code a bit? e.g.
```
const typeToRun = await getVersionToRun();
const result = {
    copybookLocation: "",
    typeToRun,
    ...
};
```

File: clients/cobol-lsp-vscode-extension/src/extension.ts [hunk 6]
```diff
@@ -75,14 +78,14 @@ async function initialize(context: vscode.ExtensionContext) {
     outputChannel,
   );
   if (maybeZowe && "futureApi" in maybeZowe) {
-    maybeZowe.futureApi.then((api) => {
+    void maybeZowe.futureApi.then((api) => {
```
Comment (slavek-kucera) [inline clients/cobol-lsp-vscode-extension/src/extension.ts:81]: Does this suppress a "dropped promise" warning or something?

File: clients/cobol-lsp-vscode-extension/src/services/Settings.ts [hunk 7]
```diff
@@ -45,63 +45,138 @@ import {
...
+const CompileOptionsConfigurationCodec = t.string;
+
+async function handleProcessorGroupConfigurationRequest<Type, Output, R>(
+  codec: t.Type<Type, Output, unknown>,
+  processorGroupLoader: (
+    requestItem: { section: string; scopeUri: string },
+    cfg: Type,
+  ) => Promise<R>,
+  item: Item,
+  result: unknown[],
```
Comment (slavek-kucera) [inline clients/cobol-lsp-vscode-extension/src/services/Settings.ts]: Should this be `R[]`?

File: clients/cobol-lsp-vscode-extension/src/services/copybook/downloader/DownloadUtil.ts [hunk 8]
```diff
@@ -182,7 +182,15 @@ export class DownloadUtil {
     return action === UNLOCK_DOWNLOAD_QUEUE_MSG;
   }
 
-  private static isInvalidCredentials(e: any) {
-    return e?.mDetails?.errorCode === 401;
+  private static isInvalidCredentials(e: unknown) {
+    return (
+      e &&
+      typeof e === "object" &&
+      "mDetails" in e &&
```
Comment (slavek-kucera) [inline clients/cobol-lsp-vscode-extension/src/services/copybook/downloader/DownloadUtil.ts]: is possible to write a `hasMember` function? Something like
```
function hasMember<U, T>(e: U, m: T): T is keyof U
```
?

File: clients/cobol-lsp-vscode-extension/src/services/reporter/TelemetryReporterImpl.ts [hunk 9]
```diff
@@ -76,8 +76,8 @@ export class TelemetryReporterImpl implements TelemetryReport {
     }
   }
 
-  public dispose(): any {
-    this.reporter.dispose();
+  public async dispose(): Promise<void> {
+    await this.reporter.dispose();
```
Comment (slavek-kucera) [inline clients/cobol-lsp-vscode-extension/src/services/reporter/TelemetryReporterImpl.ts:80]: is the dispose method really async?

File: clients/cobol-lsp-vscode-extension/src/services/util/Utils.ts [hunk 10]
```diff
@@ -49,11 +49,11 @@ export async function getExtensionApi<T>(
   }
   return {
     futureApi: new Promise((res, _) => {
-      const extAdded = vscode.extensions.onDidChange(() => {
+      const extAdded = vscode.extensions.onDidChange(async () => {
         const ext = vscode.extensions.getExtension(extName);
         if (!ext) return;
         extAdded.dispose();
-        extractApi<T>(ext, validate).then((api) => res(asAPI<T>(api)));
+        await extractApi<T>(ext, validate).then((api) => res(asAPI<T>(api)));
```
Comment (slavek-kucera) [inline clients/cobol-lsp-vscode-extension/src/services/util/Utils.ts]: This `await` does not really make much sense to me. It is supposed to resolve the `futureApi` promise.

File: clients/cobol-lsp-vscode-extension/src/test/suite/index.ts [hunk 11]
```diff
@@ -15,32 +15,34 @@
...
+import type { Config, NYC } from "nyc";
 
 export async function run(): Promise<void> {
   const sourceRoot = path.join(__dirname, "..", "..");
 
   // initialize nyc code coverage
-  // eslint-disable-next-line @typescript-eslint/no-require-imports
-  const NYC = require("nyc");
-  const nyc = new NYC({
+  const NYC = (await import("nyc")) as new (cfg: Config) => NYC;
```
Comment (slavek-kucera) [inline clients/cobol-lsp-vscode-extension/src/test/suite/index.ts:24]: We could try moving to `c8` here as well, to get rid of all of this.

File: clients/cobol-lsp-vscode-extension/src/__tests__/services/snippets/SnippetTest.spec.ts (hunk 3 above)
Comment (hohy) [reply clients/cobol-lsp-vscode-extension/src/__tests__/services/snippets/SnippetTest.spec.ts:257]: The original `position` "mock" is wrong. It is mocked as a function, but it should be an object. So, later when the `position` is used by the `getCurrentLineText` function to slice current line, it's basically doing `document.lineAt(undefined).text.slice(0, undefined).trim()`. And that returns whole first line of the document.

File: clients/cobol-lsp-vscode-extension/src/extension.ts (hunk 6 above)
Comment (hohy) [reply clients/cobol-lsp-vscode-extension/src/extension.ts:81]: Yes, eslint expects all promises to be awaited. The `void` operator is way how to mark promise as intentionally not awaited.

File: clients/cobol-lsp-vscode-extension/src/services/copybook/downloader/DownloadUtil.ts (hunk 8 above)
Comment (hohy) [reply clients/cobol-lsp-vscode-extension/src/services/copybook/downloader/DownloadUtil.ts]: Good idea 👍 Added ✅

File: clients/cobol-lsp-vscode-extension/src/services/reporter/TelemetryReporterImpl.ts (hunk 9 above)
Comment (hohy) [reply clients/cobol-lsp-vscode-extension/src/services/reporter/TelemetryReporterImpl.ts:80]: yes, it is 
image

File: clients/cobol-lsp-vscode-extension/src/test/suite/index.ts (hunk 11 above)
Comment (hohy) [reply clients/cobol-lsp-vscode-extension/src/test/suite/index.ts:24]: I would do that in separate PR.

File: clients/cobol-lsp-vscode-extension/src/__tests__/commands/SmartTabCommandTest.spec.ts (hunk 1 above)
Comment (hohy) [reply clients/cobol-lsp-vscode-extension/src/__tests__/commands/SmartTabCommandTest.spec.ts:70]: There was no real reason. As it was `undefined` in original implementation, I assumed it's not important for the test. But 0 would be probably better.

File: clients/cobol-lsp-vscode-extension/src/__mocks__/UriMock.ts [hunk 12]
```diff
@@ -6,9 +6,9 @@ export class Uri {
   }
 
   static file(str: string): Uri {
-    const p = str.replace(/\\/g, "/");
-
-    return new Uri((p[1] === ":" ? "/" : "") + p);
+    // const p = str.replace(/\\/g, "/");
+    return new Uri(str);
+    // return new Uri((p[1] === ":" ? "/" : "") + p);
```
Comment (slavek-kucera) [inline clients/cobol-lsp-vscode-extension/src/__mocks__/UriMock.ts]: cleanup?

**Review by slavek-kucera — APPROVED** [review]

**Pull request merged**